timeseries/src/autogluon/timeseries/trainer/abstract_trainer.py

```diff
@@ -878,15 +862,174 @@ def score(
+            model_set.update(self.get_minimum_model_set(model_name))
+        if len(model_set) > 1:
+            model_to_level = self._get_model_levels()
+            model_set = sorted(model_set, key=model_to_level.get)
+        logger.debug(f"Prediction order: {model_set}")
+
+        for model_name in model_set:
+            if model_name not in model_pred_dict:
+                try:
+                    predict_start_time = time.time()
+                    model_pred_dict[model_name] = self._predict_model(
+                        model=model_name,
+                        data=data,
+                        known_covariates=known_covariates,
+                        model_pred_dict=model_pred_dict,
+                    )
+                    pred_time_dict_marginal[model_name] = time.time() - predict_start_time
+                except Exception as e:
+                    if raise_exception_if_failed:
+                        raise RuntimeError(f"Model {model_name} failed to predict")
+                    else:
+                        logger.error(f"Model {model_name} failed to predict with the following exception:")
+                        logger.error(traceback.format_exc())
+                        model_pred_dict[model_name] = None
+                        pred_time_dict_marginal[model_name] = None
+        if self.cache_predictions:
+            self._save_cached_pred_dicts(
+                dataset_hash, model_pred_dict=model_pred_dict, pred_time_dict=pred_time_dict_marginal
+            )
+        pred_time_dict_total = self._get_total_pred_time_from_marginal(pred_time_dict_marginal)
+
+        final_model_pred_dict = {model_name: model_pred_dict[model_name] for model_name in model_names}
+        final_pred_time_dict_total = {model_name: pred_time_dict_total[model_name] for model_name in model_names}
+        if record_pred_time:
+            return final_model_pred_dict, final_pred_time_dict_total
+        else:
+            return final_model_pred_dict
+
+    def _get_total_pred_time_from_marginal(self, pred_time_dict_marginal: Dict[str, float]) -> Dict[str, float]:
+        pred_time_dict_total = defaultdict(float)
+        for model_name in pred_time_dict_marginal.keys():
+            for base_model in self.get_minimum_model_set(model_name):
+                if pred_time_dict_marginal[base_model] is not None:
+                    pred_time_dict_total[model_name] += pred_time_dict_marginal[base_model]
+        return dict(pred_time_dict_total)
+
+    @property
+    def _cached_predictions_filename(self) -> Path:
+        return Path(self.path) / "cached_predictions.pkl"
+
+    @staticmethod
+    def _compute_dataset_hash(
+        data: TimeSeriesDataFrame, known_covariates: Optional[TimeSeriesDataFrame] = None
+    ) -> str:
+        """Compute a unique string that identifies the time series dataset."""
+        from hashlib import md5
+
+        def hash_df(df: Optional[pd.DataFrame]) -> str:
```

Comment:
Should this live in autogluon.common? Seems generic. Probably something that Tabular could use to enhance some sanity checks.

Reply:
If standalone, would be useful to add a unit test (for example, hash a df, save it, load it, hash the loaded, check for equivalence).

Reply:
Moved to `autogluon.common.utils.uitls` & added tests to `autogluon/common/tests/unittests/test_hash_pandas_df.py`.